### ISSUE TITLE
Reduce Instance Profile Permissions for AWS Shoot Clusters

### ISSUE THREAD

**Issue by marwinski** [gardener/gardener#178]
# Story
- As user/provider I want to grant the worker nodes as few permissions as possible, so that the attack surface is as minimal as possible (against active hacking, but also accidental exposure).

# Motivation

- Security
- A cluster administrator user (which we provide by default) can access the cloud provider with the cloud profile user. The permissions can be found [here](https://github.com/gardener/gardener/blob/922cdbf79683a0eb40135c8ea96f924d0be5debd/charts/seed-terraformer/charts/aws-infra/templates/_main.tf#L318). Especially the `ec2:Describe*` reveals more information and possibly business relevant data than a kubernetes cluster operator wants to share.

# Acceptance Criteria
- [ ] Investigate whether we can be more specific with `ec2:Describe*` and reduce this (potentially only to cluster-owned resources with the tags)
- [ ] Investigate whether we can drop `ecr:*` (after a deprecation period, because it is an incompatible change and in violation of https://kubernetes.io/docs/concepts/containers/images "Kubernetes has native support for [ECR]"; Gardener doesn't need it anymore, but the customer might store and access images in/from ECR without having a proper ImagePullSecret)

## Step 1

These are the Describe methods used by the AWS cloud provider - all others are not needed. As a first step those can be mentioned explicitly in the cloud profile.

```
        DescribeInstances(request *ec2.DescribeInstancesInput) ([]*ec2.Instance, error)
        DescribeVolumes(request *ec2.DescribeVolumesInput) ([]*ec2.Volume, error)
        DescribeVolumeModifications(*ec2.DescribeVolumesModificationsInput) ([]*ec2.VolumeModification, error)
        DescribeSecurityGroups(request *ec2.DescribeSecurityGroupsInput) ([]*ec2.SecurityGroup, error)
        DescribeSubnets(*ec2.DescribeSubnetsInput) ([]*ec2.Subnet, error)
        DescribeRouteTables(request *ec2.DescribeRouteTablesInput) ([]*ec2.RouteTable, error)
        DescribeVpcs(input *ec2.DescribeVpcsInput) (*ec2.DescribeVpcsOutput, error)
        DescribeLoadBalancers(*elb.DescribeLoadBalancersInput) (*elb.DescribeLoadBalancersOutput, error)
        DescribeLoadBalancerPolicies(input *elb.DescribeLoadBalancerPoliciesInput) (*elb.DescribeLoadBalancerPoliciesOutput, error)
        DescribeLoadBalancerAttributes(*elb.DescribeLoadBalancerAttributesInput) (*elb.DescribeLoadBalancerAttributesOutput, error)
        DescribeLoadBalancers(*elbv2.DescribeLoadBalancersInput) (*elbv2.DescribeLoadBalancersOutput, error)
        DescribeLoadBalancerAttributes(*elbv2.DescribeLoadBalancerAttributesInput) (*elbv2.DescribeLoadBalancerAttributesOutput, error)
        DescribeTargetGroups(*elbv2.DescribeTargetGroupsInput) (*elbv2.DescribeTargetGroupsOutput, error)
        DescribeTargetHealth(input *elbv2.DescribeTargetHealthInput) (*elbv2.DescribeTargetHealthOutput, error)
        DescribeTargetGroupAttributes(*elbv2.DescribeTargetGroupAttributesInput) (*elbv2.DescribeTargetGroupAttributesOutput, error)
        DescribeListeners(*elbv2.DescribeListenersInput) (*elbv2.DescribeListenersOutput, error)
        DescribeAutoScalingGroups(*autoscaling.DescribeAutoScalingGroupsInput) (*autoscaling.DescribeAutoScalingGroupsOutput, error)
        DescribeKey(*kms.DescribeKeyInput) (*kms.DescribeKeyOutput, error)
        DescribeInstanceGroup(instanceGroupName string) (InstanceGroupInfo, error)

```

## Step 2:

Evaluate which of the functions are actually invoked by the kubelet code (`kubernetes/pkg/kubelet`). 

## Step 3: 

Evaluate which of the above methods might actually be problematic and reveal information. 

## Step 4: 

Reduce the permissions defined in step 1 above and test.

# Release Notes
``` user release notes
- Your worker nodes for AWS clusters have now minimal permissions allowing access to only this cluster.
```

# Definition of Done
- [ ] Knowledge is distributed: Have you spread your knowledge in pair programming/code review?
- [ ] Unit tests are provided: Have you written automated unit tests?
- [ ] Integration tests are provided: Have you written automated integration tests?
- [ ] Minimum API exposure: If you have added/changed public API, was it really necessary/is it minimal?
- [ ] Operations guide: Have you updated the operations guide about ops-relevant changes?
- [ ] User documentation: Have you updated the READMEs/docs/how-tos about user-relevant changes?

**Comment by vlerenc** [comment]
I don't understand. We narrowed that down (@rfranzke) and you checked it, so what is this about? Doing it again or have you found a problem? If so, which permission? The "implementation proposals" don't add up.

**Comment by vlerenc** [comment]
@marwinski Ping... any update here or new insights via Kinvolk?

**Comment by marwinski** [comment]
Update: we have done some investigations and here are the results in a nutshell:

- the credentials provided by the metadata service correspond to the role in the instance profile
- all other information does not appear to be critical as it does not reveal information on the account or other instances (ec2, volumes, ...)
- the metadata service appears to be read-only - no modifications appear to be possible at least on AWS
- the kubelet requires access to the metadata service

As for 2. above the only real problematic statement appears to be "ec2:Describe*" as this reveals a lot of information about the whole account. The cloud provider code for AWS appears only to use two of those (and they might not even be used from the kubelet). 

I am not sure the ecr: roles are needed anymore.

**Comment by vlerenc** [comment]
Thanks, I can work with that and have sharpened the user story, used the default template and added the release notes and DoD.

However, I am sceptical whether we should do that, simply because on both points the out-of-tree cloud providers might help us in a much safer way:
- EC2: Out-of-tree cloud providers will make permissions for the `kubelet` most likely obsolete and we might even drop them for good, so why investigate at all? Reducing the permissions is difficult (which is why we haven't done it in the past), because it requires code scanning and/or experimentation (which may accidentally miss and therefore remove permissions that are required under certain circumstances).
- ECR: While the Gardener doesn't need ECR anymore, the user may still store and access private images from ECR, so we can't reduce these permissions without risking breaking running clusters. Also, removing this, is actually in violation of https://kubernetes.io/docs/concepts/containers/images that states:
> "Kubernetes has native support for [ECR]"

**Comment by vlerenc** [comment]
@marwinski Next time please be as specific as possible, otherwise the ticket will remain in limbo until it gets more clearly spec'ed out. Thanks! In regards to out-of-tree cloud providers could you please take a look and advice. I have the feeling that we shouldn't waste too much time on the first (ECR) if out-of-tree cloud providers are coming and may not do the latter (ECR), for which I don't know what the plans are when out-of-tree cloud providers are a reality.

**Comment by marwinski** [comment]
Updated implementation proposal from issue #275.

**Comment by rfranzke** [comment]
I've evaluated to which extent we can reduce the `ec2:Describe*` action and found out that the only permission the kubelet needs is `ec2:DescribeInstances`.

Unfortunately, AWS does still not provide a way of defining resource-level permissions for all `Describe` actions [1, 2, 3].

Anyway, we will get rid the any interactions with the cloud API triggered by kubelet once we can introduce the CSI for all supported clouds (or with the extension work we are planning to do; then the AWS Botanist can independently decide what to do).

We decided to keep the `ecr:...` permissions for now.

[1] https://forums.aws.amazon.com/thread.jspa?threadID=152048
[2] https://aws.amazon.com/blogs/security/demystifying-ec2-resource-level-permissions/
[3] https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ExamplePolicies_EC2.html#iam-example-classiclink
